app/routes/_protected.verify-email.($code).tsx

```diff
@@ -1,32 +1,121 @@
-import { useLocation, useParams } from 'react-router';
+import { verifyEmail as osVerifyEmail } from '@opensecret/react';
+import type { QueryClient } from '@tanstack/react-query';
+import { useEffect } from 'react';
+import { redirect, unstable_createContext } from 'react-router';
 import { Page, PageContent } from '~/components/page';
-import { Redirect } from '~/components/redirect';
+import { LoadingScreen } from '~/features/loading/LoadingScreen';
 import { VerifyEmailForm } from '~/features/signup/verify-email-form';
-import { shouldVerifyEmail as shouldUserVerifyEmail } from '~/features/user/user';
-import { useUser } from '~/features/user/user-hooks';
-
-export default function VerifyEmail() {
-  const user = useUser();
-  const location = useLocation();
-  const { code } = useParams<{ code?: string }>();
-  const shouldVerifyEmail = shouldUserVerifyEmail(user);
-
-  if (!shouldVerifyEmail) {
-    const searchParams = new URLSearchParams(location.search);
-    const redirectTo = searchParams.get('redirectTo') || '/';
-
-    return (
-      <Redirect
-        to={{ ...location, pathname: redirectTo }}
-        logMessage="Redirecting from verify email to /"
-      />
-    );
+import { authStateQueryKey } from '~/features/user/auth';
+import {
+  type FullUser,
+  type User,
+  shouldVerifyEmail as shouldUserVerifyEmail,
+} from '~/features/user/user';
+import { userQueryKey } from '~/features/user/user-hooks';
+import { toast } from '~/hooks/use-toast';
+import { getQueryClient } from '~/query-client';
+import type { Route } from './+types/_protected.verify-email.($code)';
+
+const verifyEmailContext = unstable_createContext<FullUser>();
+
+const getUser = (queryClient: QueryClient) => {
+  const user = queryClient.getQueryData<User>([userQueryKey]);
+  if (!user) {
+    // This should never happen because route guard middleware in _protected.tsx will run first and load the user.
+    throw new Error('User not found');
+  }
+  return user;
+};
+
+const getRedirect = (request: Request) => {
+  const location = new URL(request.url);
+  const redirectTo = location.searchParams.get('redirectTo') || '/';
+  // We have to use window.location.hash because location that comes from the request does not have the hash
+  return redirect(`${redirectTo}${location.search}${window.location.hash}`);
+};
+
+const routeGuardMiddleware: Route.unstable_ClientMiddlewareFunction = async (
+  { request, context },
+  next,
+) => {
+  const queryClient = getQueryClient();
+  const user = getUser(queryClient);
```

Comment:
this is the reason I had to use this react router middleware feature which is unstable/beta (but I think will be made stable soon) atm. because this route is a child route of _protected.tsx and it needs to read the user data loaded by that route. I couldn't do that using clientLoaders because as I mentioned on the call yesterday clientLoaders run in parallel so child would attempt to read the user before the parent loads it.

middleware is different because child middleware is executed only after parent middleware calls `await next()`